mllib/src/main/scala/org/apache/spark/ml/fpm/AssociationRules.scala

```diff
@@ -0,0 +1,234 @@
+   * Default: "freq"
+   *
+   * @group param
+   */
+  final val freqCol: Param[String] = new Param[String](this, "freqCol", "frequency column name")
+
+
+  /** @group getParam */
+  @Since("2.2.0")
+  final def getFreqCol: String = $(freqCol)
+
+  /** @group setParam */
+  @Since("2.2.0")
+  def setFreqCol(value: String): this.type = set(freqCol, value)
+  setDefault(freqCol -> "freq")
+
+  /**
+   * Param for minimum confidence, range [0.0, 1.0].
+    *
+    * @group param
+   */
+  final val minConfidence: DoubleParam = new DoubleParam(this, "minConfidence", "min confidence",
+    ParamValidators.inRange(0.0, 1.0))
+
+  /** @group getParam */
+  @Since("2.2.0")
+  final def getMinConfidence: Double = $(minConfidence)
+
+  /** @group setParam */
+  @Since("2.2.0")
+  def setMinConfidence(value: Double): this.type = set(minConfidence, value)
+  setDefault(minConfidence -> 0.8)
+
+  /**
+   * Computes the association rules with confidence above [[minConfidence]].
+   *
+   * @return a DataFrame("antecedent", "consequent", "confidence") containing the association
+   *         rules.
+   */
+  @Since("2.2.0")
+  override def fit(dataset: Dataset[_]): AssociationRulesModel = {
+    val freqItemSetRdd = dataset.select($(itemsCol), $(freqCol)).rdd
+      .map(row => new FreqItemset(row.getSeq[String](0).toArray, row.getLong(1)))
+
+    val spark = SparkSession.builder().getOrCreate()
+    import spark.implicits._
+    val oldModel = new MLlibAssociationRules()
+      .setMinConfidence($(minConfidence))
+      .run(freqItemSetRdd)
+      .map(r => (r.antecedent, r.consequent, r.confidence))
+      .toDF("antecedent", "consequent", "confidence")
+    copyValues(new AssociationRulesModel(uid, oldModel).setParent(this))
+  }
+
+  @Since("2.2.0")
+  override def transformSchema(schema: StructType): StructType = {
+    SchemaUtils.checkColumnType(schema, $(itemsCol), new ArrayType(StringType, false))
+    SchemaUtils.checkColumnType(schema, $(freqCol), LongType)
+    SchemaUtils.appendColumn(schema, $(predictionCol), new ArrayType(StringType, false))
+  }
+
+  override def copy(extra: ParamMap): AssociationRules = defaultCopy(extra)
+
+}
+
+
+/**
+ * :: Experimental ::
+ * Model fitted by AssociationRules.
+ *
+ * @param associationRules AssociationRules
+ */
+@Since("2.2.0")
+@Experimental
+class AssociationRulesModel private[ml] (
+    @Since("2.2.0") override val uid: String,
+    val associationRules: DataFrame)
+  extends Model[AssociationRulesModel] with AssociationRulesParam with MLWritable {
+
+  /** @group setParam */
+  @Since("2.2.0")
+  def setItemsCol(value: String): this.type = set(itemsCol, value)
+
+  /** @group setParam */
+  @Since("2.2.0")
+  def setPredictionCol(value: String): this.type = set(predictionCol, value)
+
+  @Since("2.2.0")
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val rules = associationRules.rdd.map(r =>
+      (r.getSeq[String](0), r.getSeq[String](1))
+    ).collect()
```

Comment:
This looks like it will blow up clusters for large models.  Use a join instead of a UDF.